### PR TITLE
StorageOS attach device before volume attach

### LINKED ISSUE #69580: StorageOS volume mount succeeds even if request to mount via api fails
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
Using the StorageOS volume driver, the VolumeMount() api method is called after the volume driver has already performed the mount.  This means that the api is unable to reject the k8s mount request if prerequisites fail (e.g. it has been mounted elsewhere or the volume has been disabled).

In addition, no api call is made to attach the StorageOS device prior to the mount operation.  This does not currently cause a problem as with StorageOS all devices are available on all cluster nodes, but validation should still be done in the Attach phase so that any errors are raised in the correct phase.  This is done correctly in the StorageOS CSI driver and should be back-ported to the volume driver.

**What you expected to happen**:
PVC mount should fail if StorageOS VolumeMount() API call fails.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a PVC using the StorageOS Volume Driver.
2. Mount the volume using external tools (`storageos volume mount volname /mnt` on any node)
3. Reference the PVC in a Pod.  The PVC will get mounted, even though StorageOS knows it is mounted elsewhere.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8+ (incl. 1.12.0)
- Cloud provider or hardware configuration: any
- OS (e.g. from /etc/os-release): any linux (incl. Ubuntu 18.04)
- Kernel (e.g. `uname -a`): any linux (incl. 4.15.0-33-generic)
- Install tools: any (incl. minikube)
- Others:

/sig storage
/cc @darkowlzz

### PR DESCRIPTION
**Pull request by darkowlzz**
**What this PR does / why we need it**:
This fixes an issue in StorageOS volume plugin where volume mount succeeds even if request to mount via StorageOS API fails, by adding an `AttachDevice` API, which is called before a volume is attached. Also, updates the StorageOS API library.

**Which issue(s) this PR fixes**:
Fixes #69580

**Release note**:

```release-note
StorageOS volume plugin updated to fix an issue where volume mount succeeds even if request to mount via StorageOS API fails.
```

### CODE REVIEW

**Comment by darkowlzz** [comment]
Not sure what it failed. Some help would be appreciated.
Running it again.

/restest

**Comment by darkowlzz** [comment]
Not sure what it failed. Some help would be appreciated.
Running it again.

/retest

**Comment by croomes** [comment]
Hi @rootfs - would you mind having a look when you have time?

Thanks!

**Comment by saad-ali** [comment]
/approve

Please move the volume expand to a separate PR.
Will let @gnufied review and lgtm

/assign @gnufied

**Comment by darkowlzz** [comment]
I've removed the volume expand related changes from this PR. Will create a separate PR for that.
Thanks!

**Review by gnufied — COMMENTED** [review]
Also the title of the PR still says "Volume expansion" please remove it.

File: pkg/volume/storageos/storageos.go [hunk 1]
```diff
@@ -347,6 +349,14 @@ func (b *storageosMounter) SetUp(fsGroup *int64) error {
 		b.volNamespace = b.podNamespace
 	}
 
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.pvName, b.podNamespace, b.volName)
+
+	// Attach the device to the host.
+	if err := b.manager.AttachDevice(b, globalPDPath); err != nil {
+		klog.Errorf("Failed to attach device at %s: %s", globalPDPath, err.Error())
+		return err
```
[the remark this answers is not in the document]
Comment (darkowlzz) [reply pkg/volume/storageos/storageos.go]: Was just being consistent with the existing format in the same file and the other plugins.

File: pkg/volume/storageos/storageos.go [hunk 2]
```diff
@@ -277,6 +277,8 @@ type storageosManager interface {
 	CreateVolume(provisioner *storageosProvisioner) (*storageosVolume, error)
 	// Attaches the disk to the kubelet's host machine.
 	AttachVolume(mounter *storageosMounter) (string, error)
+	// Attaches the device to the host at a mount path.
+	AttachDevice(mounter *storageosMounter, deviceMountPath string) error
```
[the remark this answers is not in the document]
Comment (darkowlzz) [reply pkg/volume/storageos/storageos.go]: `AttachDevice` calls a StorageOS API for the virtual volumes which are created on the host and it doesn't actually attaches a device. The StorageOS virtual volume is then used by `AttachVolume` to attach the volume to the kubelet.

File: pkg/volume/storageos/storageos.go [hunk 3]
```diff
@@ -347,6 +349,14 @@ func (b *storageosMounter) SetUp(fsGroup *int64) error {
 		b.volNamespace = b.podNamespace
 	}
 
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.pvName, b.podNamespace, b.volName)
+
+	// Attach the device to the host.
+	if err := b.manager.AttachDevice(b, globalPDPath); err != nil {
```
[the remark this answers is not in the document]
Comment (darkowlzz) [reply pkg/volume/storageos/storageos.go]: I don't think there's a need for that. It just creates a virtual volume by calling a StorageOS API.

File: pkg/volume/storageos/storageos.go [hunk 4]
```diff
@@ -347,6 +349,14 @@ func (b *storageosMounter) SetUp(fsGroup *int64) error {
 		b.volNamespace = b.podNamespace
 	}
 
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.pvName, b.podNamespace, b.volName)
```
Comment (gnufied) [inline pkg/volume/storageos/storageos.go]: How does access pattern of storageOS works on multiple pods using same volume on same node. Can the volume be attached multiple times on the node with different paths?  

In general I am not sure why this volume uses pod's namespace in something that it calls "GlobalPDPath". That implies this path can only be reused if another pod within same namespace uses same volume. What happens if another pod but in different namespace uses the volume? I am not super familiar with StorageOS and hence these questions.

Comment (darkowlzz) [reply pkg/volume/storageos/storageos.go]: Yes, a volume can be used by multiple pods on the same node if they are in the same namespace as the PVC.

`b.podNamespace` looks to be incorrect there. Should be `b.volNamespace`. I'll change that.

The name variable name `globalPDPath` can be changed too. It does sounds misleading. It's just the target path where a loop device is mounted and made accessible to a pod.
I'll rename it to target path.

Comment (gnufied) [reply pkg/volume/storageos/storageos.go]: okay, once tests are fixed I will lgtm.

**Comment by darkowlzz** [comment]
Can someone help with this test failure?
It doesn't look like it's related to this change at all: `waiting for ready nodes timed out`.

**Comment by thockin** [comment]
Punting to newly minted dep-approver @apelisse !

**Comment by apelisse** [comment]
Godeps: Updating StorageOS dependency, no other new dependencies. Seems reasonable to me.

/approve

**Pull request merged**